### ISSUE TITLE
overlap removal warnings and errors, part 1

### ISSUE THREAD

**Issue by lemzwerg** [fontforge/fontforge#4184]
This is the first of a series of bug reports related to FontForge's overlap removal code.  It gets used in LilyPond using FontForge's batchmode to produce the 'Emmentaler' fonts for music notation.

[Based on sources from 2020-02-22 18:07 UTC-ML-TtfDb-D-GDK3.\
 Based on source from git with hash: af998cc28fc69ccce56ffbaf2c62aa57d2790bef\
 PythonUI_Init()\
 copyUIMethodsToBaseTable()\
 fontforge 20190801\
 build date: 2020-02-22 18:07 UTC]

Attached is a small input PFB file (`scripts.turn.pfb`) together with a small script (`scripts.turn.txt`).  Executing
```
fontforge -script scripts.turn.txt scripts.turn.pfb
```
produces the following messages.
```
Internal Error (overlap) in scripts.turn: Fixup error 2 in MergeIntersections.
Internal Error (overlap) in scripts.turn: Could not find finalm [missing newline]
```
The images below show that FontForge apparently does the right thing, so the warnings seem not necessary.
[scripts.turn.zip](https://github.com/fontforge/fontforge/files/4241121/scripts.turn.zip)
[scripts.turn.txt](https://github.com/fontforge/fontforge/files/4241126/scripts.turn.txt)
![scripts turn before](https://user-images.githubusercontent.com/1630245/75111351-201b3500-5639-11ea-9731-b927e03abb9c.png)
![scripts turn after](https://user-images.githubusercontent.com/1630245/75111354-23aebc00-5639-11ea-92c4-0b68b4f52c86.png)

**Comment by lemzwerg** [comment]
[Version: 20201107
 Based on sources from 2021-11-07 08:31 UTC-ML-TtfDb-D-GDK3.]

Still the same (harmless) problem.

**Comment by ctrlcctrlv** [comment]
Have you considered Skia, Werner? It's got some bugs, but less than `splineoverlap.c`. There's also @Logicalshift's `flo_curves::bezier::path::PathCombine`, which I haven't really used too much but I've considered switching out MFEKglif's Skia PathOps code for it and seeing if it's superior or inferior.

**Comment by lemzwerg** [comment]
No, I haven't, and it would be very inconvenient for us to use that.  The build system of LilyPond is already an incredibly large and complicated monster; adding new dependencies can easily cause cross-compilation nightmares.

Additionally, this (and other) overlap issues are bugs, and bugs should be fixed IMHO :-)

**Comment by ctrlcctrlv** [comment]
Certainly, I was just curious. FWIW, Skia has no trouble with the broken test case in #4186.

![image](https://user-images.githubusercontent.com/838783/143047636-646f9db7-531e-4a8b-9773-1fc472561804.png)

**Comment by lemzwerg** [comment]
Good to know, thanks.  BTW, for our Emmentaler font we have circumvented all overlap bugs by (slightly) rewriting the original METAFONT code.  And it seems that we have to continue with doing this because a new FontForge bug appeared, see #4859.  Sigh.

**Comment by ctrlcctrlv** [comment]
@lemzwerg I was holding this utility back, but your thread inspired me to finish it so I could actually compare flo_curves' output to Skia's. (The private build of the utility only supported Skia.)

https://github.com/MFEK/pathops

My initial observation is that flo_curves is more _correct_, but results in considerably more points, while Skia is more conservative about adding points, but has well known long-standing issues.

Here's the flo_curves output:

![image](https://user-images.githubusercontent.com/838783/143085374-05f4261c-1b48-4923-a66d-4c8a2a739bfb.png)

Now I know that you don't want to add another utility to your build, and this isn't meant as a suggestion you do so, I actually need this utility for my own font development work so it's not "wasted". The option's at least now there if you or anyone else needs it.

**Comment by lemzwerg** [comment]
Nice!
